src/compiler/types.ts

```diff
@@ -1486,6 +1494,12 @@ module ts {
     export interface StringLiteralType extends Type {
         text: string;  // Text of string literal
     }
+    
+    export interface TypeGuardType extends Type {
+        parameterName: string;
+        parameterIndex?: number; // Call expression argument

```

Comment:
If the parameterName is a name found in a binding pattern, will parameterIndex be the index of the top level binding pattern? How will you then know which binding element needs to be narrowed? Also, please ensure you have tests that try to type guard a parameter inside a binding pattern.